src/opensles/EngineOpenSLES.h

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2017 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef OBOE_ENGINE_OPENSLES_H
+#define OBOE_ENGINE_OPENSLES_H
+
+#include <atomic>
+#include <mutex>
+
+#include <SLES/OpenSLES.h>
+#include <SLES/OpenSLES_Android.h>
+
+namespace oboe {
+
+/**
+ * INTERNAL USE ONLY
+ */
+class EngineOpenSLES {
+public:
+    static EngineOpenSLES *getInstance();
+
+    SLresult open();
+
+    void close();
+
+    SLresult createOutputMix(SLObjectItf *objectItf);
+
+    SLresult createAudioPlayer(SLObjectItf *objectItf,
+                               SLDataSource *audioSource,
+                               SLDataSink *audioSink);
+    SLresult createAudioRecorder(SLObjectItf *objectItf,
+                                 SLDataSource *audioSource,
+                                 SLDataSink *audioSink);
+
+private:
```

Comment:
The constructor and the destructor need to be made private, and copy and assignment constructors deleted. Otherwise, someone could make another instance of this singleton.

See example code here: http://www.modernescpp.com/index.php/thread-safe-initialization-of-a-singleton

Reply:
Done